// jest-dom/dist/to-be-invalid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toBeInvalid = toBeInvalid;
exports.toBeValid = toBeValid;

var _jestMatcherUtils = require("jest-matcher-utils");

var _utils = require("./utils");

var FORM_TAGS = ['input', 'select', 'textarea'];

function isElementHavingAriaInvalid(element) {
  return element.hasAttribute('aria-invalid') && element.getAttribute('aria-invalid') !== 'false';
}

function isElementInvalid(element) {
  return !element.checkValidity();
}

function toBeInvalid(element) {
  var _this = this;

  (0, _utils.checkHtmlElement)(element, toBeInvalid, this);
  var isInvalid = isElementHavingAriaInvalid(element) || isElementInvalid(element);
  return {
    pass: isInvalid,
    message: function message() {
      var is = isInvalid ? 'is' : 'is not';
      return [(0, _jestMatcherUtils.matcherHint)(`${_this.isNot ? '.not' : ''}.toBeInvalid`, 'element', ''), '', `Received element ${is} currently invalid:`, `  ${(0, _jestMatcherUtils.printReceived)(element.cloneNode(false))}`].join('\n');
    }
  };
}

function toBeValid(element) {
  var _this2 = this;

  (0, _utils.checkHtmlElement)(element, toBeValid, this);
  var isValid = !isElementHavingAriaInvalid(element) && FORM_TAGS.includes((0, _utils.getTag)(element)) && !isElementInvalid(element);
  return {
    pass: isValid,
    message: function message() {
      var is = isValid ? 'is' : 'is not';
      return [(0, _jestMatcherUtils.matcherHint)(`${_this2.isNot ? '.not' : ''}.toBeValid`, 'element', ''), '', `Received element ${is} currently valid:`, `  ${(0, _jestMatcherUtils.printReceived)(element.cloneNode(false))}`].join('\n');
    }
  };
}
